Remote server integration tests documentation (#227)
mozilla/remote-settings: README.rst

Integration tests can be executed on a remote server.

To run the integration test suite, first build the integration tests container

.. code-block:: shell

    docker-compose build tests

or download a pre-built container from `Dockerhub <https://hub.docker.com/r/mozilla/remote-settings-integration-tests>`_.

Next run the tests, supplying config values as necessary. Config values are
set as environment variables provided to the Docker container. See
``tests/conftest.py`` for descriptions of all of the config options that are
available.

Note that the tests assume that the server has the ``attachments``,
``changes``, ``history``, and ``signer`` plugins enabled. It may optionally
have the ``email`` plugin installed.

To have the tests bootstrap themselves (i.e. when ``SKIP_SERVER_SETUP=false``):

> - a user account should available with the ability to create users, buckets, and

Reviewer: ```suggestion
- The credentials passed in ``SETUP_AUTH`` should have the permission to create users, buckets, and
```

Author: [d57f3d5](https://github.com/mozilla/remote-settings/pull/227/commits/d57f3d50673d86a2da6792c7f0c9ec4085bc5056)

  collections

> - the account should also be able to assign users to groups

Reviewer: This permission does not really exist in itself. It the user has the permission create/write buckets, then it can write groups

Author: [d57f3d5](https://github.com/mozilla/remote-settings/pull/227/commits/d57f3d50673d86a2da6792c7f0c9ec4085bc5056)

- the credentials of this user should be supplied to the container

If the tests should not bootstrap themselves and instead use resources already
available on the server (i.e. when ``SKIP_SERVER_SETUP=true``):

> - There should a bucket and collection available.

Reviewer: ```suggestion
- The ``BUCKET`` (default: ``main-workspace``) and ``COLLECTION`` (default: ``product-integrity``) must exist
```

Author: [d57f3d5](https://github.com/mozilla/remote-settings/pull/227/commits/d57f3d50673d86a2da6792c7f0c9ec4085bc5056)

- There should be two users available
  - one user should be added to the ``editor`` group of the available
    collection

>   - the other should be added to the ``reviewer`` group

Reviewer: The group names are plural (eg. ``product-integrity-editors``)

Author: [8cf424f](https://github.com/mozilla/remote-settings/pull/227/commits/8cf424f97bbb7e0e859ab881b832649f06a111b1)

- the names of the bucket, collection, and user credentials should be supplied

>   as environment variables to the container

Reviewer: I think you could rewrite these three points by mentioning the env var names

Author: [d57f3d5](https://github.com/mozilla/remote-settings/pull/227/commits/d57f3d50673d86a2da6792c7f0c9ec4085bc5056)

Integration tests can be executed on a remote server.

To run the integration test suite, first build the integration tests container

.. code-block:: shell

    docker-compose build tests

or download a pre-built container from `Dockerhub <https://hub.docker.com/r/mozilla/remote-settings-integration-tests>`_.

Next run the tests, supplying config values as necessary. Config values are
set as environment variables provided to the Docker container. See
``tests/conftest.py`` for descriptions of all of the config options that are
available.

Note that the tests assume that the server has the ``attachments``,
``changes``, ``history``, and ``signer`` plugins enabled. It may optionally
have the ``email`` plugin installed.

To have the tests bootstrap themselves (i.e. when ``SKIP_SERVER_SETUP=false``),
the credentials passed in ``SETUP_AUTH`` should have the permission to create 

> users, buckets, and collections.

Reviewer: Shall we provide an example of auth? (eg. `SETUP_AUTH=admin:s3cr3t`  or `SETUP_AUTH="Bearer xzy123"`

Author: [9dfaa00](https://github.com/mozilla/remote-settings/pull/227/commits/9dfaa0047a11a8ec109fda94c996ea9b794b753c)

If the tests should not bootstrap themselves and instead use resources already
available on the server (i.e. when ``SKIP_SERVER_SETUP=true``):

> - There should a bucket and collection available

Reviewer: ```suggestion
- There should be a bucket and collection available
```

  
  - the bucket, if not specified by the ``BUCKET`` config option, should be named ``main-workspace``
  - the collection, if not specified by the ``COLLECTION`` config option, should be named ``product-integrity``

- There should be two users available

  - one user should be added to the ``editors`` group of the available collection
  - the other should be added to the ``reviewers`` group of the available collection
  - the credentials of these users should be passed in the ``EDITOR_AUTH`` and
    ``REVIEWER_AUTH`` config options respectively

> Running integration tests on the Remote Settings dev server should look something like:

Reviewer: nitpick?
```suggestion
Running integration tests on the Remote Settings DEV server should look something like:
```

.. code-block:: shell

    docker run --rm \
        --env SERVER=https://settings.dev.mozaws.net/v1 \

>         --env MAIL_DIR="" \

Reviewer: You don't want to keep the comment about disabling tests?

Author: I do, but I realized that comment block doesn't actually do anything. Unless you know a better way, I don't think there's a great way to include comments inline with shell commands.

Author: Yeah, probably better to leave this in.

        --env SKIP_SERVER_SETUP=true \
        --env TO_REVIEW_ENABLED=false \
        --env EDITOR_AUTH=<credentials available in 1Password> \

>         --env REVIEWER_AUTH=<credentials available in 1Password> \

Reviewer: Shall we give a hint about format or username here?

```suggestion
        --env REVIEWER_AUTH=editor:<password available in 1Password> \
```

Author: [9dfaa00](https://github.com/mozilla/remote-settings/pull/227/commits/9dfaa0047a11a8ec109fda94c996ea9b794b753c)
